### DISCUSSION TITLE
How to install the rook ceph cluster in a seLinux enforcing env (not an openshift env)

### DISCUSSION THREAD

**Discussion by abiao610** [rook/rook#9559] [Q&A]
Hi, 

I used to install the rook ceph cluster to envs (1 master + 3 workers; k8s v1.21.7; crio v1.22; rook v1.7) with seLinux disabled, but these days I have to install it to target envs with seLinux enabled (with status of 'enforcing').  

Since the tatget env is not an openshift env, then I will not able to do "kubectl create -f cluster/examples/kubernetes/ceph/scc.yaml" for the k8s resource type of 'scc' is not available in my target env. 

I googled around, and after I added "securityContext: privileged: true" to operator.yaml, I can see the pod of " rook-ceph-operator-xxxxxxxxx" running. Then I can't go any further, since I have no idea how to add "securityContext: privileged: true" to the rest of the containers. After I do "kubectl create -f cluster.yaml", I can see the following 2 pods keep 'CrashLoopBackOff', and also see related audit denied log from the k8s node where these 2 pods hosted.

rook-ceph-detect-version-mpscx
rook-ceph-csi-detect-version-9v5df

Please help and thanks a lot in advance.

**Comment by leseb** [comment]
Hi, you can set `ROOK_HOSTPATH_REQUIRES_PRIVILEGED` to true in the `rook-ceph-operator-config` configmap.

**Reply by abiao610** [reply]
Thank u for response!

I found "rook-ceph-operator-config" configmap dose not have "ROOK_HOSTPATH_REQUIRES_PRIVILEGED", and I manually add this "ROOK_HOSTPATH_REQUIRES_PRIVILEGED" and set it to "true".  But still failed.

**Reply by leseb** [reply]
What's the failure?

**Reply by abiao610** [reply]
Details of the failures:
1. the 2 'detect' pods keep status of 'Init:CrashLoopBackOff'
2. describe the 'detect' pods and see 'Error from server (BadRequest): container "cmd-reporter" in pod "rook-ceph-detect-version-29g8d" is waiting to start: PodInitializing'
3. log into the 2 k8s nodes where the 2 'detect' pods locate, and check with the audit logsby using cmd of "cat /var/log/audit/audit.log | grep avc", and seeing the following 2 messages repeating for 6 or 7 times.

on the node where pod 'rook-ceph-csi-detect-version-zfcvk' locates:
type=AVC msg=audit(1642050135.408:4509): avc:  denied  { transition } for  pid=576164 comm="3" path="/tini" dev="overlay" ino=1029499 scontext=system_u:system_r:unconfined_service_t:s0 tcontext=system_u:system_r:container_t:s0:c484,c989 tclass=process permissive=0

on the node where pod 'rook-ceph-detect-version-29g8d' locates:
type=AVC msg=audit(1642050133.282:5791): avc:  denied  { ioctl } for  pid=551368 comm="tini" path="/dev/null" dev="tmpfs" ino=6 ioctlcmd=0x5410 scontext=system_u:system_r:container_t:s0:c204,c377 tcontext=system_u:object_r:container_file_t:s0:c204,c377 tclass=chr_file permissive=0

**Reply by abiao610** [reply]
BTW, I am working on openstack instances, not bare metal envs. I am not sure it matters or not, and I can install successfully when seLinux disabled.

**Comment by Madhu-1** [comment]
@abiao610 the above audit logs are not related to cephcsi. as we don't use tini in cephcsi image.

**Reply by abiao610** [reply]
hi Madhu:

I think so, and this 'tini' is a binary file from the pod of 'rook-ceph-operator-75c9f9cf9c-nr6jf' for rook release 1.7.

I think 'tini' try to do something from node #1 (where the pod of 'rook-ceph-operator-75c9f9cf9c-nr6jf' locates) to some other nodes (where the 2 'detect' pods locate), but it can't for seLinux is enforcing on the other nodes.

**Comment by KKonak** [comment]
I am also trying to get rook ceph running with SELinux enforcing. Did you find a solution to this?

**Reply by fskale** [reply]
I use SELINUX MLS in enforcing mode. I developed a container policy over the years and it works well.
NetApp Trident must ran as SPC_T which is the only permissive type besides container_runtime_t.
I changed the deployment of the mons and added:
`securityContext:
        seccompProfile:
          type: RuntimeDefault
        seLinuxOptions:
          level: "s0"
          type: "spc_t"`
 It then deployed mon-b but still got stuck at "chown /var/lib/rook".
 I then switched to permissive mode, but id didn't change anything.
 Any help would be greatly appreciated !

**Reply by fskale** [reply]
I found it myself:
1.) operator.yaml
add:
`
seccompProfile:
              type: RuntimeDefault
            seLinuxOptions:
              level: "s0"
              type: "spc_t"
`
and
set `ROOK_HOSTPATH_REQUIRES_PRIVILEGE` to `true` !



2.) csi-operator.yaml
add the same under "securityContext"

Now you have both pods running as spc_t

spc_t must be a permissive type !

`
system_u:system_r:spc_t:s0      2016     2452448  0.1  0.1 1308408 96808 ?       Ssl  09:14   0:00 /usr/local/bin/rook ceph operator
system_u:system_r:spc_t:s0      65532    2455248  0.1  0.0 1278700 54680 ?       Ssl  09:19   0:00 /manager --leader-elect
`

3.) Create cluster 

4.) Check the ceph status
`
kubectl -n rook-ceph exec -it deploy/rook-ceph-tools -- ceph mon stat
e3: 3 mons at {a=[v2:xxxxx:3300/0,v1:xxxxx:6789/0],b=[v2:xxxx:3300/0,v1:1xxxx:6789/0],c=[v2:xxxx:3300/0,v1:xxxx:6789/0]} removed_ranks: {} disallowed_leaders: {}, election epoch 12, leader 0 a, quorum 0,1,2 a,b,c
`

**Reply by KKonak** [reply]
Been awhile, but I ended up creating my own SELinux policy that inherits container_t and used gatekeeper to mutate all deployments/jobs/pods with my custom context. This allowed me to run Rook/Ceph not privileged. Glad you figured it out

**Reply by fskale** [reply]
> Been awhile, but I ended up creating my own SELinux policy that inherits container_t and used gatekeeper to mutate all deployments/jobs/pods with my custom context. This allowed me to run Rook/Ceph not privileged. Glad you figured it out

The problem with elevated container_t is, that it can also read/write up levels.  When using the redhat container policy, container_rumtime_t starts the pod which will then be running under container_t (C1, C2).

The core services as well as some services which will need kernel access must run as spc_t.
I'll remove the spc_t permissiveness once i have all the rejects filtered and added to the custom policy.
It would be great to have a project, like secured Kubernetes, where are the "geeks" can input their policies which will then be greatly valueable for all the security enthusiasts…
